### PR TITLE
Issue #14084: Remove checkerframework violation for DefaultConfigurat…

### PR DESCRIPTION
**Pull request by Rohanraj123**
related #14084

### CODE REVIEW

**Comment by nrmancuso** [comment]
@Rohanraj123 let's get existing PRs out of draft mode before opening more: https://github.com/checkstyle/checkstyle/pulls/Rohanraj123

File: config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
Comment (nrmancuso) [inline config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml:147]: Annotation must be placed on the method that has this line as it's return statement. Please take some time to read through the checker framework docs to understand this tool: https://checkerframework.org/manual/

**Comment by Rohanraj123** [comment]
> @Rohanraj123 let's get existing PRs out of draft mode before opening more: https://github.com/checkstyle/checkstyle/pulls/Rohanraj123

Not gonna open any more before these !!

File: config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
Comment (Rohanraj123) [reply config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml:147]: > Please take some time to read through the checker framework docs to understand this tool: https://checkerframework.org/manual/

On it sir.

**Comment by Rohanraj123** [comment]
Reasoning before this change : 
Heuristics for Handling Common Cases:
The Nullness Checker uses certain heuristics to handle common cases when determining the nullness type of the returned array:
If the passed array has a length of 0 or is created with size 0 (new String[0]), the returned array component type is @NonNull.
And This is the case with us too.

Reason for SuppressWarning("nullness")
Due to the complexities in analyzing toArray method calls, the explanation suggests that you may need to use @SuppressWarnings("nullness") annotation when overriding the toArray method. 

Document reference : https://checkerframework.org/manual/#null-defaults

**Comment by Rohanraj123** [comment]
> Reason for SuppressWarning("nullness")

Okay After using this SupressionWarnings, I saw it was not allowed. As such. But I think this is the only way to resolve this issue as per the checker framework documentation. Can you please suggest me some way via which I can use suppression warnings in DefaultConfiguration.java file.

File: src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:105]: Why are we converting a list to an array and streaming it, instead of directly streaming the array?

Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java]: If not converting the list to array, yields new pitest error.

**Comment by romani** [comment]
@Rohanraj123, do you have time to continue now?
Or we can close it for now and you will reopen it when you have time to continue it.

**Comment by Rohanraj123** [comment]
> @Rohanraj123, do you have time to continue now? Or we can close it for now and you will reopen it when you have time to continue it.

I will contiinue my work from here.

**Comment by Rohanraj123** [comment]
@romani I got the problem now. 
Basically as per the checker framework error. It says that method return type is `NonNull` whereas return expression is `Nullable`.
This is because how checker framework analyzes the `toArray`() method.
Official source : 
`The nullness semantics of [Collection.toArray(T[])](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#toArray(T[])) cannot be captured by just the nullness type system`

returned array is of type @NonNull if : 
1 . collections type is Nonnull
2. size(passed array) <= size(collections)

Since we have children list that is converted into array which is non null because its a  initialized list of configuration
and same goes with passed array. 

Therefor as per my understanding Two possible solutions are here :
1. Use suppression warnings
2. using dynamically created array parameter. 

As I did in this commit . Can you suggest any other method if this ain't work

**Comment by Rohanraj123** [comment]
@romani This PR is ready for review !

Changes : 
To remove violations of checker framework. 
--> I just removed the constant predefined due to the lack of analysis of checker framework in toArray() framework. It sometimes becomes overly cautious. 
--> So I passed the dynamically created array in toArray() method to guarantee the checker framework that param array is not null.

File: src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:104]: This means you found defect in checkerframefork tool.
Please create issue on it, close this PR, make new PR with comments over xml suppression blocks with comment `<-- until #xxxxxx -->`

**Review by romani — CHANGES_REQUESTED** [review]
Item

**Comment by Rohanraj123** [comment]
@ArrayLen() annotation can only work when Constant value checker is configured. And After looking pom.xml file, constant value checker is configured. 
so now. adding command line option `Alint=trustArrayLenZero` in compilerargs should process annotation @ArrayLen() .

File: src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:39]: Restore this line

File: src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:42]: Put annotation above, not on same line

**Review by romani — CHANGES_REQUESTED** [review]
CI is red

File: src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java
Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:42]: @romani putting annotation above giving me violation in `checker-formatter` that it should be either before or after of the `private static final `

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:43]: Please put it before

Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:43]: done!

Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java]: > @romani putting annotation above giving me violation in `checker-formatter` that it should be before type and after modifiers.

**Comment by Rohanraj123** [comment]
Implemtation yet done : 
1. Used @ArrayLen() annotation at correct position as per the checker framework.
2. To process that annotation and verify . I added the command line option : `-Alint=trustArrayLenZero`
3. To maintain consistency among checkers profiles i added this command line arg.

Issue still : 
Nullchecker still doesnt process the annotation .
@romani

**Comment by Rohanraj123** [comment]
@romani Can you confirm me whether `Alint` command is configured correctly or not. 
Because if that is okay then something is wrong in  the CF tool. Because it doesnt recoganises the ArrayLen() annotation.

I have tried all the possible ways.

File: pom.xml
Comment (romani) [inline pom.xml:2540]: Try to put new lines before `</compilerArguments>`

**Review by romani — CHANGES_REQUESTED** [review]
Item

File: pom.xml
Comment (Rohanraj123) [reply pom.xml:2541]: done!

Comment (romani) [reply pom.xml:2541]: Not done

Comment (romani) [reply pom.xml:2541]:
```
<compilerArguments>
                <Xlint>all</Xlint>
...
```

Comment (Rohanraj123) [reply pom.xml]: Nope It still aint work !

Comment (romani) [reply pom.xml]: Let's ask for help from checkerframefork team, please create issue or discussion in their repository

Comment (Rohanraj123) [reply pom.xml]: Done : https://github.com/typetools/checker-framework/issues/6702

Comment (Rohanraj123) [reply pom.xml]: Screenshot from 2024-07-09 06-49-34

Screenshot from 2024-07-09 06-49-31

Screenshot from 2024-07-09 06-49-16

Screenshot from 2024-07-09 06-49-09

@romani  These are the results i found when I actually configured checker framework in one of the demo project. 

It says Unsupported Command line option : Alint=trustArrayLenZero

Comment (romani) [reply pom.xml]: Please share this test project in GitHub and show it to it maintainer, to let them suggest how to configure it.

Comment (romani) [reply pom.xml]: There's also `<compilerArgs combine.children="append">` try to put it there

Comment (Rohanraj123) [reply pom.xml]: > There's also `<compilerArgs combine.children="append">` try to put it there

Its already there !

**Comment by romani** [comment]
for now this PR is blocked by https://github.com/typetools/checker-framework/issues/6702

**Comment by Rohanraj123** [comment]
@romani This is the only way to place annotation I have tried all the ways in my demo project. 
But CI gives syntax error. Should I work on resolving the  syntax error or should we ditch the idea of using annotation?

**Comment by romani** [comment]
Yes, please make code compile able.

**Comment by Rohanraj123** [comment]
@romani After looking into the problem. I found that plexus is generating this issue because not being able to generate the meta data due to syntax incompatibility. Now The only way is to change the place of annotation should not be there or excluding pexus from our codebase and using something else that can just be compatible with any syntax type. 

Do you have somework around in your mind that we can implement ? Please suggest!

**Comment by Rohanraj123** [comment]
One more last thing i wanna confirm from you. I am thinking of using this workaround that I preprocess  the files and temporarily remove the annotations during meta data generation and then again restore it afterwards. I can write scripts for this purpose and and integerate in pom.xml file.

**Comment by romani** [comment]
No, add/Remove is super hack, looks like we need to go deeper in plugin that failing on such syntax.

**Comment by romani** [comment]
we lost energy in this PR, I closing it, we will be back to it one day in future.

**Pull request closed without merging**
